Expression plots/tables: tumors only and tumor vs normal

komalsrathi opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section

#### What scientific question is your analysis addressing?

1. Create `tumor vs normal` and `tumor only` expression plots, with log2(tpm + 1) on the y-axis, as follows:

* cohort + cancer_group tumor only plots: For each `cohort + cancer_group`, plot all tumors.
* cancer_group tumor only plots: For each `cancer_group`, plot all tumors.
* cohort + cancer_group tumor vs normal plots: For each cancer_group within a cohort, plot `cohort + cancer_group` vs GTEx subgroups.

2. Export tables (.tsv) of the data per plot with the following fields: `gene`, `mean`, `median`, `sd` and `x-axis-labels`.

3. Export metadata files to use in JSON files with the following fields: `gene` (gene name), `plot_type` (either `tumors_only` or `tumor_vs_normal`), `cohort` (`_` separated cohort name), `analysis_type` (either `cohort_cancer_group_level` or `cancer_group_level`), `uuid`, `plot_fname` (plot filename) and `table_fname` (table filename)

#### What was your approach?

1. Combine `CBTN` and `PNOC` into one cohort: `PBTA`.
2. Filter `cancer_subgroup` or `gtex_subgroup` with number of samples < 5 

#### What GitHub issue does your pull request address?

https://github.com/PediatricOpenTargets/ticket-tracker/issues/38

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

Filenames and content of `results/*.tsv` and `plots/*.png`
Filename and content of `results/metadata.tsv`

#### Is there anything that you want to discuss further?

Currently only generating outputs for top 10 genes for each comparison for the initial review. I have pushed them under `results` and `plots` only for the purpose of this review. Once the plots, filenames and data formats are finalized, I can remove those from the repo and push them directly to s3. 

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

N/A

### Results

#### What types of results are included (e.g., table, figure)?

1. `plots/*.png`: expression boxplots with `log2(tpm + 1)` as y-axis
Filename format: `{gene_name}_{cohort_name}_{analysis_type}_{uuid}.png` 

2. `results/*.tsv`: corresponding data for expression boxplots
Filename format: `{gene_name}_{cohort_name}_{analysis_type}_{uuid}.tsv`

3. `results/metadata.tsv`: mapping of filename and metadata for all comparisons 

#### What is your summary of the results?

N/A

#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

<!-- IF YOUR PULL REQUEST IS A DATA RELEASE, PLEASE REMOVE THE [HTML COMMENT TAG](https://html.com/tags/comment-tag/) FROM THE SECTION BELOW AND COMPLETE THE CHECKLIST-->

<!-- 
### Data Release Checklist
- [ ] Is the table in doc/data-file-descriptions.md up to date?
- [ ] Is doc/data-format.md up to date?
- [ ] Is doc/release-notes.md up to date?
- [ ] Is download-data.sh up to date?
- [ ] Was download-data.sh tested and did it complete without error?
-->

komalsrathi commented:
@kgaonkar6  question: do I have to add this to the `.circleci/config.yml`, I didn't add mine because I was searching for the newer modules for e.g. the ones that @logstar opened but could not find any in it - am I missing something?

kgaonkar6 commented:
We have not yet set up test data so haven't been maintaining the CI. I'm checking if the modules runs in docker while I review the PR.

jharenza commented:
Hi @komalsrathi - I think the plotting looks good, but let's hold off for generating all plots because I noticed some things.

- GMKF_NBL has 209 not 208 samples per your earlier comment about lack of histology, so we will want to be sure that is included
- We may be able to get rid of MT genes to reduce the number of plots we make. Checking with @taylordm 
- We have a few edits to PBTA for cancer group coming in v6, and I noticed Cranio is split here with adamantinomatous being one group- we may want to collapse.

jharenza commented:
hey @komalsrathi - one thing I noticed is one missing plot type, which is the `cancer_group` vs GTEX subgroup. So, instead of a plot named `MT-ATP6_GMKF_GTEx_cohort_cancer_group_level_422afd0ad46a11ebbeabf2189859b6c1.png`, it would be something like `MT-ATP6_Neuroblastoma_GTEx_cohort_cancer_group_level_422afd0ad46a11ebbeabf2189859b6c1.png`. Can you add this?

komalsrathi commented:
Gotcha, sure. I'll update soon.

komalsrathi commented:
To be consistent, I will change the names for `cohort + cancer_group` tumor vs normal plots to include the `cancer_group` name as well. For e.g. filename of the plot below is `MT-RNR2_PBTA_GTEx_cohort_cancer_group_level_1d905642d46b11ebbeabf2189859b6c1.png` so instead of that it would be `MT-RNR2_PBTA_Craniopharyngioma_GTEx_cohort_cancer_group_level_1d905642d46b11ebbeabf2189859b6c1.png`
![MT-RNR2_PBTA_GTEx_cohort_cancer_group_level_1d905642d46b11ebbeabf2189859b6c1](https://user-images.githubusercontent.com/7000615/123322371-358cc180-d502-11eb-975a-ed5fa63213a9.png)

jharenza commented:
sounds good!

komalsrathi commented:
@jharenza I need clarification on this: 
the UUID generated in the [tumor plots](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/expression-plots/analyses/tumor-gtex-plots/util/tumor_plot.R#L49) and [tumor vs normal plots](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/expression-plots/analyses/tumor-gtex-plots/util/tumor_vs_normal_plot.R#L64) functions is not reproducible so each time we run the script, it is going to change the UUID suffix for a given plot/table. 

Also, the `tumor vs normal` function is called once (it would be twice after we add the above missing plot type) and `tumor vs tumor` function is called twice in the [bash script](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/expression-plots/analyses/tumor-gtex-plots/run-tumor-gtex-plots.sh), so there is a chance that the same `UUID suffix` is generated for two different plots.

jharenza commented:
for 
>Also, the tumor vs normal function is called once (it would be twice after we add the above missing plot type) and tumor vs tumor function is called twice in the bash script, so there is a chance that the same UUID suffix is generated for two different plots.

Can you generate a list of the number of UUIDs needed, pull from them from function 1 and then remove before function 2?

I have to check on the other point at how much of an issue that will become... hopefully we can generate these once as soon as all of the data are analyzed.

jharenza commented:
ok @komalsrathi regarding point 1 about irreproducibility of UUIDs, @taylordm suggested we go back to human readable naming

komalsrathi commented:
@jharenza sounds good - update soon!

kgaonkar6 commented on analyses/tumor-gtex-plots/01-tumor-gtex-plots.R:
```diff
@@ -0,0 +1,96 @@
+
+# match expression matrix to histologies file
+expr_mat <- expr_mat %>%
+  select(all_of(hist_file$Kids_First_Biospecimen_ID))
```
The version of tidyselect(0.2.5) we have in the docker doesn't seem to support `all_of()`

kgaonkar6 commented on analyses/tumor-gtex-plots/01-tumor-gtex-plots.R:
```diff
@@ -0,0 +1,96 @@
+# just do top 10 genes as a test, we can remove this later
+n_genes <- 10
+expr_mat <- expr_mat %>%
+  slice_head(n = n_genes)
```
do we need splice_head() here or a `head()` would do? I don't think we have `tidyft` installed in the docker

kgaonkar6 reviewed (commented):
Thanks @komalsrathi code looks good to me! Just some minor questions regarding the functions which are not available in the version of tidyverse we have in our docker. 

I didn't review the plots in detail.. I'd suggest we can do 1 gene as example run, thoughts @jharenza @komalsrathi?  The top 10 genes still results in ~500 files which we would not be able to review in detail if an update is done to the module.

komalsrathi commented:
@kgaonkar6 Sounds good - making these changes

kgaonkar6 reviewed (approved):
Thanks for the updates!

jharenza commented:
Hi @komalsrathi ! We have some additional updates/guidance from the 8 am call this morning re: plots. We are also awaiting some information from NCI about the tables (they may want them condensed into larger tables, but not sure yet).

Plot updates requested:
- Get rid of cohort prefix on X-axis but retain cohort information in the title
- Use TPM instead of log2 TPM on y-axis
- Convert (disease_N) to disease (N = #) in x-axis
- If possible/easy fix, clean title to read similar to "Gene GMKF NBL vs. GTEX" (gene, cohort, cancer_group vs GTEX) or Gene Neuroblastoma vs. GTEX" (gene, cancer_group vs GTEX)
- Alphabetize GTEX tissues on x-axis, but keep the cancer group as the first group

Thank you!

komalsrathi commented:
Sure, these should be quick so I'll update soon

komalsrathi commented:
@jharenza just one question regarding: `Get rid of cohort prefix on X-axis but retain cohort information in the title ` if we have TARGET Neuroblastoma and GMKF Neuroblastoma, that would be confusing right - if we remove the cohort name from x-axis? We would have to color by cohort to differentiate

jharenza commented:
Oh sorry, I meant that comment specific to the cohort level plots. For the cancer level, disregard the cohort info.

komalsrathi commented:
> Oh sorry, I meant that comment specific to the cohort level plots. For the cancer level, disregard the cohort info.

That plot above is a pan-cancer cohort + cancer group level plot. For cancer level, I don't use the cohort info (see plot 2) in the x-axis:

1) PBTA +  GMKF Cohort + Cancer level: 
![MT-CO3_PBTA_GMKF_cohort_cancer_group_level](https://user-images.githubusercontent.com/7000615/123989697-24770100-d997-11eb-926c-8b0819c61a09.png)

2) PBTA +  GMKF Cancer level:
![MT-CO3_PBTA_GMKF_cancer_group_level](https://user-images.githubusercontent.com/7000615/123989691-2345d400-d997-11eb-80a7-cdfc7caf4099.png)

For plot 1) if we add Neuroblastoma from TARGET, we will need cohort info in the x-axis.

jharenza commented:
Ah, I responded too quickly, yes - so for the 4 plots:
1. cohort+cancer_group vs GTEX, no cohort info in the x-axis, as this is one cohort per plot, but retain in the title - eg: `GMKF Neuroblastoma vs GTEX`
2. cancer_group vs GTEX, no cohort info in the x-axis since they are combined anyway, and no cohort info in the title - eg: `Neuroblastoma vs GTEX`
3. pan-cancer by cohort, you're right, we will need the cohort in the x-axis. Since they want these alphabetized, can you add cohort at the end, so for eg: `Neuroblastoma, GMKF (N = #)` or `Neuroblastoma, TARGET (N = #)`. The title of these can just be `Gene Expression across cohorts` (without having to name any cohorts)
4. pan-cancer by cancer_group, no cohort in the x-axis. The title of these can just be `Gene Expression across cancers` 

Does that sound better?

komalsrathi commented:
Done, for a quick review, I have mentioned all types of unique plots/result files in the [README](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/expression-plots/analyses/tumor-gtex-plots#module-structure)

jharenza commented:
You're so fast! These look great. Could you do me a favor since this gene is so highly expressed, just make the plots for GPC2? Then, I think the plots are ready to go. We can hold on generating all, though, until we have the rest of the TARGET RNA harmonized and released.

komalsrathi commented:
Makes sense - updated.

jharenza commented:
perfect, thank you!

sangeetashukla commented:
@jharenza @komalsrathi Is this ready for me to go ahead and prepare json files?

jharenza commented:
> @jharenza @komalsrathi Is this ready for me to go ahead and prepare json files? 

No, hold off on this one because we may turn it into a shiny app- haven't gotten clear direction yet.

jharenza reviewed (changes requested):
@komalsrathi now that #34 is in, will you update this using the v6 histologies file still with GPC2 (we condensed a few cancer groups)? Now, we also have PBTA as cohort, so you can remove the code combining CBTN and PNOC. Still TBD about what we will do with the tables.

komalsrathi commented:
@jharenza in `v6` there are 30 samples that have `cohort == "PBTA"` and `cancer_group == NA` - should I add a filter to remove those before plotting?

jharenza commented:
> @jharenza in `v6` there are 30 samples that have `cohort == "PBTA"` and `cancer_group == NA` - should I add a filter to remove those before plotting? Are the `NAs` intentional? 

Yes, those were benign/other tumors so only use which samples are present in cancer_group. Thanks for checking!

komalsrathi commented:
Ok, thanks for clarifying - updated to v6.

jharenza commented:
hi @komalsrathi - if you can add https://github.com/PediatricOpenTargets/ticket-tracker/issues/119 to this PR, we can update and close, and then add https://github.com/PediatricOpenTargets/ticket-tracker/issues/110 and annotation later. Thanks!

jharenza commented on analyses/tumor-gtex-plots/README.md:
```diff
@@ -0,0 +1,124 @@
+	--hist_file=HIST_FILE
+		histologies file (.tsv)
+
+	--cohort_list=COHORT_LIST
+		comma separated list of cohorts
```
```suggestion
	--cohort_list=COHORT_LIST
		comma separated list of cohorts
```

For this, we are going to use `all_cohorts` if we have >1 cohort, so can you update the code and output table to reflect that?

komalsrathi replied:
Actually I kept it this way so that we can use either `--cohort_list 'GMKF, PBTA, GTEx'` for tumor vs normal plots and `--cohort_list 'GMKF, PBTA'` for tumors only plots. Because for tumors only plots, we are not really using all cohorts.

jharenza reviewed (changes requested):
Hi @komalsrathi - this looks good in general. Few change requests below:

1. Can you rename the tables to have this prefix: `tumor_normal_gtex_plots_` or `pan_cancer_plots_`.
2. `cancer_group_level.tsv` table first 46 rows are missing the `cancer_group`. I noticed these are referring to the pan-cancer plots not the tumor/normal plots. Can you separate those plots out since they are two different plot types per se. So, we would have `tumor_normal_gtex_plots_cancer_group_level.tsv` and `pan_cancer_plots_cancer_group_level.tsv` tables for `cancer_group level. Does that make sense? In the `cancer_group` column, you would add each cancer by row. This would actually make 4 tables instead of 2 - I forgot about the pan-cancer comparisons before. 
3. Once that is done, can you create the tables to include _all_ genes? 
Thanks!

komalsrathi commented:
1) tsv files have been split into 4 for each comparison type (pan-cancer, tumor vs normal) + analysis_type (cohort + cancer_group, cancer_group only) combo
2) `pan_cancer_plots_` tsv files have `cancer_group == x_labels` and `tumor_normal_gtex_plots_` tsv files have reference cancer group in `cancer_group`

Can you check and let me know if this is ok or do I need to make more changes - still have doubts about using `all_cohorts` as we are not using all the cohorts for pan-cancer plots.

jharenza commented:
Hi!

>pan_cancer_plots_ tsv files have cancer_group == x_labels and tumor_normal_gtex_plots_ tsv files have reference cancer group in cancer_group

`cancer_group` has to be the exact match to the cancer group in order for the EFO/MONDO annotation to match later - can you update this to remove the `(N = ###)`

Since @logstar has created an [annotator](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/long-format-table-utils), you will no longer need to include the columns for `efo_code`, `mondo_code`, `uberon_code` - these will be added with the annotator (I will submit a separate issue for adding this). You can also either remove the ENSG id column and I can add it to the next issue with annotation or add the annotation using this file in the data release: `ensg-hugo-rmtl-v1-mapping.tsv`

>Can you check and let me know if this is ok or do I need to make more changes - still have doubts about using all_cohorts as we are not using all the cohorts (i.e. GTEx is not being used) for pan-cancer plots.

Using `all_cohorts` implies all cancer cohorts - I see what you are saying here, but it will be all disease cohorts when we represent the tables in the OT portal, so we have to update this to match all other tables please.

In your table `tumor_normal_gtex_plots_cohort_cancer_group_level.tsv`, I am seeing multiple cohorts in the `cohort column`, but these should be single disease cohorts. Cohort here again means disease/cancer cohort. But let me ask about the rows for GTEX and designating them as such - maybe we just designate the comparator row as the cohort or `all_cohorts`.

komalsrathi commented:
Ok I see, so the `cohort` column would be `all_cohorts` for `pan_cancer_plots_` because those include all tumor cohorts and for `tumor_normal_gtex_plots_cohort_cancer_group_level`, it would be either `PBTA` or `GMKF` (i.e. the reference tumor cohort)?

jharenza commented on Dockerfile:
```diff
@@ -402,9 +402,17 @@ RUN ./install_bioc.r \
 # Package for querying gene IDs and symbols
 RUN ./install_bioc.r \
     mygene
+>>>>>>> origin/dev
```
looks like a merge conflict got introduced here

jharenza reviewed (changes requested):
Hi @komalsrathi - thanks for the updates - I think this is almost there. 

Can you make the following updates:
- To both tables:
  - rename the `gene` column as `Gene_symbol` for compatibility with the annotator [here](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/long-format-table-utils).
  - rename `cancer_group` column to `Disease` for compatibility with the annotator module
  - Add `Gene_Ensembl_ID` using the mapping file `ensg-hugo-rmtl-mapping.tsv` in the v7 data release 
  - rename the `cohort` column name as `Dataset`

To the tumor_normal plot tables, we may also need to add a separate column for `GTEx_tissue_subgroup` - cc-ing @logstar here as he figures out the annotator for this. 

Thanks!

komalsrathi commented:
Thanks, I am making these changes - assuming that I should hold off on adding `GTEx_tissue_subgroup`?

jharenza commented:
> Thanks, I am making these changes - assuming that I should hold off on adding `GTEx_tissue_subgroup`?

sorry yes, please add as another separate column since that will be annotated with UBERON

komalsrathi commented:
@jharenza this is ready for review - I haven't added `GTEx_tissue_subgroup` because my understanding is it will be annotated/added with UBERON. Or did you mean create a new column but keep it `NA` (just like plot_api)?

jharenza commented:
> @jharenza this is ready for review - I haven't added `GTEx_tissue_subgroup` because my understanding is it will be annotated/added with UBERON. Or did you mean create a new column but keep it `NA` (just like plot_api)?

I meant to add it now, but we can leave it blank.

Do you want to handle annotation/JSONL conversion in another PR? It entails running this annotator (https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/long-format-table-utils) on your file, converting all `NA` to blanks, and then converting to JSONL and gzipping the file.

jharenza commented:
>Add `Gene_Ensembl_ID` using the mapping file ensg-hugo-rmtl-mapping.tsv in the v7 data release

It looks like this column header needs an update for the annotator to work.

komalsrathi commented:
> > @jharenza this is ready for review - I haven't added `GTEx_tissue_subgroup` because my understanding is it will be annotated/added with UBERON. Or did you mean create a new column but keep it `NA` (just like plot_api)?
> 
> I meant to add it now, but we can leave it blank.
> 
> Do you want to handle annotation/JSONL conversion in another PR? It entails running this annotator (https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/long-format-table-utils) on your file, converting all `NA` to blanks, and then converting to JSONL and gzipping the file.

No, I think I will try to add it to this one because this is still WIP.

komalsrathi commented:
@logstar does the annotator require that I be in VPN or am I missing something? I tried: 

```
Rscript --vanilla ../../analyses/long-format-table-utils/annotator/annotator-api.R -r -v -c MONDO,RMTL,EFO -i results/pan_cancer_plots_cohort_cancer_group_level.tsv -o results/pan_cancer_plots_cohort_cancer_group_level.tsv
```

But it does nothing. cc: @jharenza

logstar commented:
> @logstar does the annotator require that I be in VPN or am I missing something? I tried:
> 
> ```
> Rscript --vanilla ../../analyses/long-format-table-utils/annotator/annotator-api.R -r -v -c MONDO,RMTL,EFO -i results/pan_cancer_plots_cohort_cancer_group_level.tsv -o results/pan_cancer_plots_cohort_cancer_group_level.tsv
> ```
> 
> But it does nothing. cc: @jharenza

The `annotator-cli.R` should be used for command line interface.

komalsrathi commented:
> The `annotator-cli.R` should be used for command line interface.

awesome, thanks!

logstar commented:
> > The `annotator-cli.R` should be used for command line interface.
> 
> awesome, thanks!

you're welcome!

komalsrathi commented:
> Do you want to handle annotation/JSONL conversion in another PR? It entails running this annotator (https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/long-format-table-utils) on your file, converting all `NA` to blanks, and then converting to JSONL and gzipping the file.

I have added the annotator call to this module but have not converted to JSONL and gzipped the files. Do we have a module for it?

jharenza commented:
No, but you can look at lines 1024+ here in the R code here for JSON conversion https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/snv-frequencies and then the bash script for jsonl conversion.

komalsrathi commented:
I think the above commit takes care of everything that was discussed - please let me know if anything is missing.

jharenza commented:
Hi @komalsrathi - In `pan_cancer_plots_cohort_cancer_group_level.tsv`, I think that this file should not have ` all_cohorts`, but have the Dataset (PBTA, GMKF, TARGET) in that column. In `pan_cancer_plots_cancer_group_level.tsv` and `tumor_normal_gtex_plots_cancer_group_level.tsv` only have `Neuroblastoma (N = 211)`, but should be higher now. `tumor_normal_gtex_plots_cohort_cancer_group_level.tsv` doesn't have the TARGET cohort. This data was added in v7 - can you add that?

komalsrathi commented:
> In `pan_cancer_plots_cancer_group_level.tsv` and `tumor_normal_gtex_plots_cancer_group_level.tsv` only have `Neuroblastoma (N = 211)`, but should be higher now. `tumor_normal_gtex_plots_cohort_cancer_group_level.tsv` doesn't have the TARGET cohort. This data was added in v7 - can you add that?

Ok yes, I need to change the bash script to add TARGET - will do now. 

> In `pan_cancer_plots_cohort_cancer_group_level.tsv`, I think that this file should not have ` all_cohorts`, but have the Dataset (PBTA, GMKF, TARGET) in that column. 

Earlier I had the individual cohort names in all tsv files and I thought that was the idea but I think I was told to keep `all_cohorts` because that means all tumor cohorts and people were not considering the normal cohort (see https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/29#issuecomment-885071954) ? Just letting you know that pan-cancer plots always have all tumor cohorts per plot so we would also have to revert back from `all_cohorts` in `pan_cancer_plots_cancer_group_level.tsv`. 

Does that mean that both tumor vs normal analysis tsv files i.e. `tumor_normal_gtex_plots_cancer_group_level.tsv` and `tumor_normal_gtex_plots_cohort_cancer_group_level.tsv` should also have the list of cohorts mentioned?

jharenza commented:
Ah, sorry for the confusion - only the `cancer_group` files should have `all_cohorts`, as those Ns are all combined. In the files which are cohort + cancer_group and you have to denote cohort, you should use the cohort for that dataset row.

komalsrathi commented:
Ok makes sense. Thanks!

komalsrathi commented:
Ok I think that issue is fixed - added TARGET to the bash script and fixed the Dataset column in cohort + cancer_group tsv files.

jharenza commented:
Hi @komalsrathi - `pan_cancer_plots_cohort_cancer_group_level.tsv` has `PBTA, GMKF, TARGET` in each row, but this should be only one of the cohorts, whichever cohort is represented in `x_labels` column. You have this right in `tumor_normal_gtex_plots_cohort_cancer_group_level.tsv`. However, in `tumor_normal_gtex_plots_cohort_cancer_group_level.tsv`, under `x_labels`, you are missing the cohort.

You can also remove the `plot_api` column since @jonkiky will be adding that information.

Once you update these few things, I think we are ready to expand the code to the full list of genes.

komalsrathi commented:
@jharenza I also don't have the cohort name in the x-axis in the tumor-normal cohort level plots e.g. 
![GPC2_PBTA_CNS-Embryonal-tumor_vs_GTEx_cohort_cancer_group_level](https://user-images.githubusercontent.com/7000615/128253924-65b38e85-3375-4f2c-9df4-7a6a6d338174.png)

So I need to add the cohort name in x-axis too?

jharenza commented:
>So I need to add the cohort name in x-axis too? According to this: #29 (comment) we don't want the cohort name in x-labels in the plot but we want cohort name in x-labels in the tsv file?

Ah good point - I was reviewing the tables without looking at the plots and JMM said he would rather have that in the title of the plot. Ok, disregard that change. I think the table context is a bit harder when you look at it alone, but it will be downloaded from a plot anyway.

komalsrathi commented:
Updated the tsv files (think I got it right this time 😄)

jharenza commented:
These look good to me @komalsrathi! Can you generate the TSV files to include all gene data? (Can you do that separate of generating all plots?)

komalsrathi commented:
> These look good to me @komalsrathi! Can you generate the TSV files to include all gene data? (Can you do that separate of generating all plots?)

Sure, I will work on it now and update you when finished.

komalsrathi commented:
@jharenza started running this at 1 PM and it is still running, I will check back tomorrow and update here on the estimated time.

komalsrathi commented:
Both `pan_cancer_*.tsv` were generated within 1-2 hours. But `tumor_normal_gtex_plots_cohort_cancer_group_level.tsv` started shortly after 2:15 PM yesterday and still running. I monitored for a minute just now and observed ~22-25 genes processing per minute. So it should complete around 2-3 PM today. Estimating the same time for `tumor_normal_gtex_plots_cancer_group_level.tsv` which will start after this is finished (based on the order in `run-tumor-gtex-plots.sh`). So, should have all tsv files ready by tomorrow PM.  

Just a quick note, I replaced `plyr::d_ply` with a `for-loop` as it was not benefitting in terms of time taken and in fact was taking much longer for the tumor-normal data. This can probably be made faster if we use `plyr::d_ply` with multiple cores (`.parallel = TRUE` and along with library `doMC`) run on some server. I am running on my local machine so I avoid using multiple cores as it usually greatly affects my other jobs too. In terms of size:

```
117.3 MB results/pan_cancer_plots_cancer_group_level.tsv
  114 MB results/pan_cancer_plots_cohort_cancer_group_level.tsv
 3.89 GB results/tumor_normal_gtex_plots_cohort_cancer_group_level.tsv (and growing)
```

komalsrathi commented:
@jharenza I have the `.tsv` files but when I try to convert to jsonl.gz for the tumor vs normal my local machine's memory gets exhausted. Any way I can get access to a server where I can just upload the `.tsv` files and do the conversion from `.tsv` to `jsonl.gz`?

komalsrathi commented:
This is the limiting step: https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/expression-plots/analyses/tumor-gtex-plots/02-annotate-and-zip.R#L25-L28 so cannot use jsonlite::write_json to convert TSV to JSON in R for large files (in our case ~5 GB). Will explore alternatives in a bit.

jharenza commented:
> This is the limiting step: https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/expression-plots/analyses/tumor-gtex-plots/02-annotate-and-zip.R#L25-L28 so cannot use `jsonlite::write_json` to convert TSV to JSON in R for large files (in our case ~8.5 GB because the tumor-normal input `.tsv` files are 5GB but after running through the annotator they grow to ~8.5GB). Will explore alternatives in a bit.
> 
> FYI, does not work on servers as well because of R-specific memory issues:
> 
> ```
> Error in collapse(tmp, inner = FALSE, indent = indent) :
>   R character strings are limited to 2^31-1 bytes
> ```

Do you mean on an R server or an EC2 instance? Can this be run in CAVATICA? cc @kgaonkar6 and @migbro or @yuankunzhu for assistance and cc @logstar as an FYI

komalsrathi commented:
> > This is the limiting step: https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/expression-plots/analyses/tumor-gtex-plots/02-annotate-and-zip.R#L25-L28 so cannot use `jsonlite::write_json` to convert TSV to JSON in R for large files (in our case ~8.5 GB because the tumor-normal input `.tsv` files are 5GB but after running through the annotator they grow to ~8.5GB). Will explore alternatives in a bit.
> > FYI, does not work on servers as well because of R-specific memory issues:
> > ```
> > Error in collapse(tmp, inner = FALSE, indent = indent) :
> >   R character strings are limited to 2^31-1 bytes
> > ```
> 
> Do you mean on an R server or an EC2 instance? Can this be run in CAVATICA? cc @kgaonkar6 and @migbro or @yuankunzhu for assistance and cc @logstar as an FYI

No, I don't think so. On ANY server, we will have to use "streaming" from tsv and to JSON. Please give me some time to figure this out before pulling others in. Thanks.

jharenza commented:
>No, I don't think so. On ANY server, we will have to use "streaming" from tsv and to JSON. Please give me some time to figure this out before pulling others in. Thanks. I'll bring up this issue in the 3 pm call today.

Ok great, and no problem. cc-d those since we have a similar issue with another module (DeSeq2), which is requiring us to build a separate CAVATICA app for the module (was run on isilon first) and something that we may end up migrating towards with more modules as we continue to add data.

komalsrathi commented:
@jharenza I have updated the [code](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/expression-plots/analyses/tumor-gtex-plots/02-chunkwise-annotate-and-zip.R) and it works now by reading input data (.tsv) into chunks of 1 million rows, converting to JSON and then JSONL, followed by combining all JSONL files and gzipping the final JSONL file. Added this to README as well. 

Q1: Should I remove old plots and results from the repo and add `plots/` and `results/` to `.gitignore` now that we are generating it for all the genes and not feasible to upload on github? 

Q2: Where should I upload the final .tsv and .jsonl.gz files? 

Final size of the files:

```
 15M	results/pan_cancer_plots_cancer_group_level.jsonl.gz
144M	results/pan_cancer_plots_cancer_group_level.tsv
 14M	results/pan_cancer_plots_cohort_cancer_group_level.jsonl.gz
144M	results/pan_cancer_plots_cohort_cancer_group_level.tsv
321M	results/tumor_normal_gtex_plots_cancer_group_level.jsonl.gz
7.0G	results/tumor_normal_gtex_plots_cancer_group_level.tsv
336M	results/tumor_normal_gtex_plots_cohort_cancer_group_level.jsonl.gz
6.9G	results/tumor_normal_gtex_plots_cohort_cancer_group_level.tsv
```

jharenza commented:
ok great! I created a folder here for you - let me know if you cannot upload: `s3://kf-openaccess-us-east-1-prd-pbta/open-targets/analyses/tumor-gtex-plots/`

jharenza reviewed (approved):
going to get this merged if everything is good on your end now @komalsrathi?

komalsrathi commented:
> going to get this merged if everything is good on your end now @komalsrathi?

yes, I uploaded the jsonl.gz files on s3 and box. the two tumor-normal tsv files are getting slowly uploaded to s3 so good to merge.
